lib/connection_tester.rb

```diff
@@ -0,0 +1,289 @@
+        result.ssl_status = false
+      when HTTPFailure
+        result.reachable = true
+      when NodeInfoFailure
+        result.diaspora_version = ""
+      end
+    end
+  end
+
+  # @raise [AddressFailure] if the specified url is not http(s)
+  def initialize(url, result)
+    @url ||= url
+    @result ||= result
+    @uri ||= URI.parse(@url)
+    raise AddressFailure,
+          "invalid protocol: '#{@uri.scheme.upcase}'" unless a_http?(@uri)
+
+    result.hostname = @uri.host
+  end
+
+  # Perform the DNS query, the IP address will be stored in the result
+  # @raise [DNSFailure] cause by a failure to resolve or a timeout
+  def resolve
+    with_dns_resolver do |dns|
+      addr = dns.getaddress(@uri.host)
+      @result.ip = addr.to_s
+    end
+  rescue Resolv::ResolvError, Resolv::ResolvTimeout => e
+    raise DNSFailure, "'#{@uri.host}' - #{e.message}"
+  end
+
+  # Perform a HTTP GET request to determine the following information
+  #  * is the host reachable
+  #  * is port 80/443 open
+  #  * is the SSL certificate valid (only on HTTPS)
+  #  * does the server return a successful HTTP status code
+  # (can't do a HEAD request, since that's not a defined route in the app)
+  # follow up to 3 redirects by default
+  #
+  # @param [Hash] request options
+  # @param [Result]
+  # @return [Integer] HTTP status code
+  def request
+    with_http_connection do |http|
+      response = capture_response_time { http.get("/") }
+      handle_http_response(response)
+    end
+  rescue Faraday::ConnectionFailed, Faraday::TimeoutError => e
+    raise NetFailure, e.message
+  rescue Faraday::SSLError => e
+    raise SSLFailure, e.message
+  rescue ArgumentError, FaradayMiddleware::RedirectLimitReached, Faraday::Error::ClientError => e
+    raise HTTPFailure, e.message
+  end
+
+  # Try to find out the version of the other servers software.
+  # Assuming the server speaks nodeinfo
+  #
+  # @return [String] version string
+  def nodeinfo(result)
+    url = "#{uses_ssl? ? 'https' : 'http'}://#{@hostname}"
+    opts = ConnectionTester.request_opts_for_url(url)
+    port =  opts.delete(:port)
+
+    Net::HTTP.start(@hostname, port, nil, nil, nil, opts) do |http|
+      resp = http.get(NODEINFO_FRAGMENT)
+      handle_nodeinfo_error(resp)
+
+      node_resp = http.get(find_nodeinfo_url(resp.body))
+      handle_nodeinfo_error(resp)
+
+      find_diaspora_version(node_resp.body, result)
+    end
+  rescue KeyError, JSON::JSONError
+    raise NodeInfoFailure, e.message
+  end
+
+  private
+
+  def with_http_connection
+    @http ||= Faraday.new(@url) do |c|
+      c.use Faraday::Response::RaiseError
+      c.use FaradayMiddleware::FollowRedirects, limit: 3
+      c.adapter(*faraday_adapter)
+      c.headers[:user_agent] = "diaspora-connection-tester"
+      c.options.timeout = 12
+      c.options.open_timeout = 6
+    end
+    yield(@http) if block_given?
+  end
+
+  # hook for unit tests
+  def faraday_adapter
+    [:net_http]
```

Comment:
Well, the point of me suggesting faraday was also to use the same adapter as we use everywhere else. Else figuring out why the connection tester works but say webfinger or post fetching doesn't or vice versa will get rather tiresome.